features/steps/thens/alignment.py

```diff
@@ -77,11 +79,108 @@ def count_segments(logic, representation):
+    k, v = exp_type.keys(), exp_type.values()
+
+    def fn_exactly():
+        if obs_type in v:
+            return True
+
+    if "OneOf" in k:
+        # there is more than one valid representation type for this logic segment PredefinedType
+        if obs_type in exp_type["OneOf"]:
+            return True
+    elif "Exactly" in k:
+        return fn_exactly()
+    elif "multiple" in k:
+        opts = exp_type["multiple"]
+        for opt in opts:
+            k, v = opt.keys(), opt.values()
+            if "Exactly" in k:
+                return fn_exactly()
+
+    return False
+
+
+def check_segment_geometry_types(expected_types: List[Dict], observed_types: List[str]) -> List[bool]:
+    """
+    Used in ALA003 to confirm agreement of segment geometry types between business logic
+    and geometry representation.
+
+    :param expected_types: The expected segment geometry types based on the segments in the alignment business logic
+    :param observed_types: The observed segment geometry types based on the segment(s) in the shape representation
+    """
+    is_valid = list()
+
+    # confirm that both lists have the same number of items
+    if len(expected_types) != len(observed_types):
+        return [False]
+    else:
+        for expected_type, observed_type in zip(expected_types, observed_types):
+            is_valid.append(check_segment_geometry_type(expected_type, observed_type))
+
+        return is_valid
+
+
+def pretty_print_expected_geometry_type(spec: Dict, pretty: List[str]) -> List[str]:
+    """
+    Format the expected geometry types
+    """
+
+    pairs = [(k, v) for (k, v) in spec.items()]
+
+    def fn_exactly():
+        pretty.append(v)
+
+    for k, v in pairs:
+        match k:
+            case "Exactly":
+                fn_exactly()
+            case "OneOf":
+                opts = spec[k]
+                pretty.append(" or ".join(opts))
+            case "multiple":
+                opts = spec[k]
+                for o in opts:
+                    k, v = o.keys(), o.values()
+                    if k == "Exactly":
+                        fn_exactly()
+            case _:
+                pass
+
+    return pretty
+
+
+def pretty_print_expected_geometry_types(exp: List[Dict]) -> str:
+    """
+    Format the expected list of geometry types
+    """
+    pretty = list()
+
+    for d in exp:
+        pretty = pretty_print_expected_geometry_type(spec=d, pretty=pretty)
+
+    return ", ".join(pretty)
+
+
+def ala003_activation_inst(inst, context) -> Union[ifcopenshell.entity_instance | None]:
+    """
+    Used in ALA003 as reverse traversal of graph to locate the correct business logic entity
+    """
+    for candidate in context._stack[2]["instances"]:
+        if candidate is None:
+            return None
+        else:
+            for rep in candidate.Representations:
+                for item in rep.Items:
+                    if item.id() == inst[0][0].id():
```

Comment:
```suggestion
                    if item.id() == inst.id():
```
This returns, for testfile pass-ala003-representation_only.ifc, an instance of IfcAlignmentSegment for the first iteration ..